### ISSUE TITLE
server: record heap profile when under memory pressure, streamline submission process

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#26222]
We've had multiple user issues that were related to nodes running out of memory. Unless we can reproduce the specific issue in our testing, we often fail to resolve the matter.

This is a tooling problem -- we're perfectly capable of recording and looking at heap profiles and figuring out what's wrong. But in practice, we rarely get a heap profile when it matters because a) it's too late - the node already crashed and now it won't happen again b) it's complicated to communicate the instructions to users.
The recent introduction of the pprof-ui endpoints may have helped, but it doesn't solve a) nor b) convincingly because the displayed information isn't easily sent our way.

I have the following suggestions:

## Automatically record heap profiles when memory usage is high.

I think a decent indicator is that RSS is within some threshold of available system memory, something that should be checked every ~10s. (While we're here, the node should also gossip a `crdb_internal.gossip_alert` when the threshold fires). The profiles can go to a subdirectory in the `logs` directory. Each filename is of the form `heap_<date>_<score>` where `<score>` is higher the closer to threshold we were. This allows keeping only the worst N heap profiles over time which should be interesting.

## Streamline the submission of heap profiles

Find a way to detect that there was an OOM. For starters, we check whether there are any auto-collected heap profiles. Additionally, we may want to check for a marker that is written only on successful shutdown, but that's probably not important. Either way, when we think there is something to report, show a log message that gives users a command to submit these heap profiles to us OR, even better, automatically submit them (subject to the data collection cluster vars). We'd need to set up a sink (i.e. some write-only GCE bucket) that we monitor; perhaps we can emit a sentry event that points at the file in the sink to streamline this on our end.
Additionally, we may want to offer a `./cockroach debug` command that collects a heap profile and submits it. This is useful to help users that complain about high memory usage when it isn't high enough to cause OOMs.

@jordanlewis I think there's some starter project material here -- let me know if you want to pick some of this up for future new arrivals and I'm happy to help scope.

**Comment by tschottdorf** [comment]
also cc @andreimatei who cares about making debugging easier.

**Comment by petermattis** [comment]
> I think a decent indicator is that RSS is within some threshold of available system memory, something that should be checked every ~10s.

I recall that we've had difficult reliable determining system memory in some containerized environments. Another option would be go grab a heap profile whenever the RSS has grown by more than some delta (e.g. 1GiB).

**Comment by a-robinson** [comment]
> I recall that we've had difficult reliable determining system memory in some containerized environments

The bigger issue is with CPU. It's true that if the container is on a machine with a lot of other containers and hasn't had an explicit memory limit put on it that it may be OOM-killed well before it's using all the machine's memory, but that's true in non-containerized environments as well. I don't think we need to worry about any differences when running in containers for this.

**Comment by bdarnell** [comment]
> I recall that we've had difficult reliable determining system memory in some containerized environments

It took us several iterations to refine this, but I think we're now getting a pretty good estimate of system memory in most cases (certainly enough for this use case). When it's wrong, it's an overestimate rather than an underestimate, so it would be no worse than the status quo.

**Comment by tschottdorf** [comment]
The difficulties of estimating the available memory aside (@petermattis suggestion might be a good heuristic too), any concerns with doing this? I would do this one sooner rather than later as I hope it will pay good dividends.

**Comment by bdarnell** [comment]
Yeah, the RSS growth heuristic is a good one. Both the google-perftools c++ heap profiler (`HEAP_PROFILE_INUSE_INTERVAL`) and jemalloc's (`prof_gdump`) have options for this.

**Comment by petermattis** [comment]
I have no concerns about doing this. At the very least, we should enable the generation of the heap profiles which is a trivial amount of work (see the code in `cli/start.go:initMemProfile` and the code in `server/status/runtime.go`).

**Comment by bdarnell** [comment]
While we're touching `initMemProfile`, it's come to my attention that if `--log-dir` is not explicitly specified, this writes its profiles into the process's current directory instead of picking up the default log directory (`$STORE0/logs`)

**Comment by tschottdorf** [comment]
@Nishant9, would you be able to take on the generation of the heap profiles using the RSS growth heuristic?

**Comment by Nishant9** [comment]
@tschottdorf I can take this up, but I am bot able to assign this to myself.

**Comment by Nishant9** [comment]
@tschottdorf I can take this up, but I am not able to assign this to myself.

**Comment by tschottdorf** [comment]
@Nishant9 after you accept the invite at https://github.com/cockroachdb/cockroach/invitations, you should be able to. Either way, the issue is yours!

**Comment by Nishant9** [comment]
> It took us several iterations to refine this, but I think we're now getting a pretty good estimate of system memory in most cases (certainly enough for this use case). When it's wrong, it's an overestimate rather than an underestimate, so it would be no worse than the status quo.
@bdarnell Can you point me to some place in the code where that estimate is computed/used?

**Comment by bdarnell** [comment]
@Nishant9 It's computed in `server/config.go``GetTotalMemory`. It's used in `CreateEngines` when the cache and other parameters are given as fractions.

**Comment by tschottdorf** [comment]
We are collecting heap profiles automatically now. What we *don't* do is collect them in `debug zip`.
